File: old-projects/vector-defence-silverlight-port/src/Particles/ExplosionParticle.ts
import { Sprite } from "../Sprite";
import { Game } from "../Game";
import { Util } from "../Util";

export class ExplosionParticle extends Sprite {
  private readonly size: number;
  private readonly color: string;
  private readonly burnDownSpeed: number;

  constructor(x: number, y: number, particleSize: number, particleColor: string, burnDownPerFrame: number, speed = Util.randomInRange(2, 7), offset = Util.randomInRange(4, 6)) {
    super();
    this.burnDownSpeed = burnDownPerFrame;
    this.size = particleSize;
    this.color = particleColor;
    const angle = Util.randomInRange(-Math.PI, Math.PI);
    const cos = Math.cos(angle);
    const sin = Math.sin(angle);
    this.dx = speed * cos;
    this.dy = speed * sin;
    this.x = x + (offset * cos);
    this.y = y + (offset * sin);
  }

  override animate(): void {
    const slowDownFactor = 1 - (0.04 * Game.currentFrameMultiplier);
    this.dx *= slowDownFactor;
    this.dy *= slowDownFactor;
    this.updatePosition();

    if (this.isOutsideBounds) {
      this.markedForRemoval = true;
    } else {
      this.alpha -= this.burnDownSpeed * Game.currentFrameMultiplier;
      if (this.alpha <= 0) {
        this.markedForRemoval = true;
      }
    }
  }

  override render(context: CanvasRenderingContext2D): void {
    context.save();
    context.globalAlpha = Math.max(0, this.alpha);
    context.fillStyle = this.color;
    context.fillRect(this.x - (this.size / 2), this.y - (this.size / 2), this.size, this.size);
    context.restore();
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Particles/ExplosionParticle.cs
using System;
using System.Windows.Media;
using System.Windows.Shapes;

namespace VectorDefenceSL.Particles
{
    public class ExplosionParticle : Sprite
    {
        private readonly double size;
        private readonly Brush fill;
        private readonly double burnDownSpeed;

        public ExplosionParticle(double x, double y, double particleSize, Color particleColor, double burnDownPerFrame, double speed, double offset)
        {
            burnDownSpeed = burnDownPerFrame;
            size = particleSize;
            
            var angle = Util.RandomInRange(-Math.PI, Math.PI);
            var cos = Math.Cos(angle);
            var sin = Math.Sin(angle);

            DX = speed * cos;
            DY = speed * sin;

            X = x + (offset * cos);
            Y = y + (offset * sin);

            fill = new SolidColorBrush(particleColor);
            mainElement = new Rectangle
            {
                Fill = fill,
                Width = size,
                Height = size,
                IsHitTestVisible = false,
            };
            Game.AddElement(mainElement);
        }

        public ExplosionParticle(double x, double y, double particleSize, Color particleColor, double burnDownPerFrame)
            : this(x, y, particleSize, particleColor, burnDownPerFrame, Util.RandomInRange(2, 7), Util.RandomInRange(4, 6))
        {
        }

        public override void Animate()
        {
            double slowDownFactor = 1 - (.04 * Game.CurrentFrameMultiplier);
            DX *= slowDownFactor;
            DY *= slowDownFactor;
            UpdatePosition();

            if (IsOutsideBounds)
            {
                MarkedForRemoval = true;
            }
            else
            {
                alpha -= burnDownSpeed * Game.CurrentFrameMultiplier; // alpha channel
                if (alpha <= 0)
                {
                    MarkedForRemoval = true;
                }
            }
        }

        public override void Render()
        {
            fill.Opacity = alpha;
            double x = X - (size / 2);
            double y = Y - (size / 2);
            Move(x, y);
        }
    }
}
